File: ts_src/testTemplate/handler.spec.ts
// testTemplate/handler.test.ts
import { run } from "./handler";
import * as utils from "../tslib/utils";

jest.mock("../tslib/utils", () => {
    const originalUtils = jest.requireActual("../tslib/utils");
    return {
        __esModule: true,
        ...originalUtils,
        fetchTrends: jest.fn(() => ({ data: [{ datapoints: [] }] })),
    };
});
const mockedUtils = jest.mocked(utils);
test("doesn't barf", async () => {
    const event = {
        body: '{"pointName":"KlarmanHall.Elec.Solar.PowerScout3037/kW_System","timeStamp":"2022-10-05T23:58:47.390Z"}',
    };
    const result = await run(event);
    expect(result).toHaveProperty("statusCode");
    expect(result.body).toMatch("missing");
});

test("handles 400 no data", async () => {
    mockedUtils.fetchTrends.mockRejectedValue({
        response: { status: 400, data: { error: "No data" } },
        message: "whatever",
    });
    const event = { body: '{"pointName":"foo"}' };
    const result = await run(event);
    expect(result).toHaveProperty("statusCode");
    expect(result.body).toMatch("no data");
});

test("barfs when it should", async () => {
    mockedUtils.fetchTrends.mockRejectedValue({
        response: { status: 400 },
        message: "qwerty",
    });
    const event = { body: '{"pointName":"foo"}' };
    const result = await run(event);
    expect(result).toHaveProperty("statusCode");
    expect(result.body).toMatch("qwerty");
});

File: py_src/python_template/handler_spec.py
from python_template.handler import run
from requests.exceptions import HTTPError
from requests.models import Response


def test_nobarf(mocker):
    event = {
        "body": {
            "pointName": "KlarmanHall.Elec.Solar.PowerScout3037/kW_System",
            "timeStamp": "2022-10-05T23:58:47.390Z",
        }
    }
    mocker.patch(
        "python_template.handler.fetch_trends",
        return_value={"data": [{"datapoints": []}]},
    )
    result = run(event, None)
    assert "statusCode" in result
    assert "missing" in result.get("body")


def test_handle400(mocker):
    r = Response()
    r.status_code = 400
    r._content = b'{"error":"No data"}'
    mocker.patch(
        "python_template.handler.fetch_trends", side_effect=HTTPError(response=r)
    )
    event = {"body": {"pointName": "foo"}}
    result = run(event, None)
    assert "statusCode" in result
    assert "no data" in result.get("body")


def test_barf(mocker):
    r = Response()
    r.status_code = 400
    r._content = b"qwerty"
    mocker.patch(
        "python_template.handler.fetch_trends", side_effect=HTTPError(response=r)
    )
    event = {"body": {"pointName": "foo"}}
    result = run(event, None)
    assert "statusCode" in result
    assert "qwerty" in result.get("body")